### DISCUSSION TITLE
Full text search in mysql

### DISCUSSION THREAD

**Discussion by wiseaidev** [sqlalchemy/sqlalchemy#8684] [Usage Questions] [answered]
Hey everyone. I am trying to implement a full-text search behavior across multiple columns as described [here](https://docs.singlestore.com/managed-service/en/reference/sql-reference/data-definition-language-ddl/create-table.html#fulltext-behavior). Therefore, I tried to look up the SQLAlchemy docs about this topic to implement it in [the app](https://github.com/wiseaidev/fastapi-singlestore-backend) I am currently building. However, after searching for [`full text search`](https://docs.sqlalchemy.org/en/14/search.html?q=full+text+search&check_keywords=yes&area=default#), the following results showed up in return:

![Screenshot from 2022-10-20 14-01-52](https://user-images.githubusercontent.com/62179149/196931923-59710a4f-a374-40b3-9438-83901117d762.png)

Hence, my guess is that it is only available on `sqllite` and `postgresql`. Not quite sure. Awaiting a response from maintainers. Thanks for considering this discussion.

Edit: I am looking for something similar to [TSVector](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#full-text-search) but in mysql.

**Comment by zzzeek** [comment, marked as answer]
MySQL has [match](https://dev.mysql.com/doc/refman/8.0/en/fulltext-search.html)

the built in [match()](https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=column+match#sqlalchemy.schema.Column.match) method will work on MySQL.  

There is also a more full featured variant for MySQL:  https://docs.sqlalchemy.org/en/14/dialects/mysql.html#sqlalchemy.dialects.mysql.match

**Reply by wiseaidev** [reply]
Hey @zzzeek, thanks for the response. Yeah, that's what i tried first, but in the case of SingleStore(Remote MySQL server), it doesn't allow matching if the column is not marked as `FULLTEXT` as [documented here](https://docs.singlestore.com/managed-service/en/reference/sql-reference/data-definition-language-ddl/create-table.html#fulltext-behavior). Do you know how to add this keyword to the column defined in the table, something like:

```python
    first_name: str = Column("FULLTEXT", String(20), index=True)
```

If the column is not marked as so, the remote server will complain about the following:

```python
pymysql.err.OperationalError: (1191, "Leaf Error (node-XXX): Can't find FULLTEXT index matching the column list")
```

Any idea how to solve this issue?

Edit: i need to find a way to tell the SQLAlchemy engine to translate the column model into something like:

```sql
CREATE TABLE users (
    first_name VARCHAR(200),
    FULLTEXT (first_name)
);
```

Is this possible with SQLAlchemy?

**Reply by zzzeek** [reply]
You can do that on an Index for MySQL, see https://docs.sqlalchemy.org/en/14/dialects/mysql.html#index-prefixes

**Reply by wiseaidev** [reply]
Oh, thanks! However, the problem now is that I am using the declarative ORM. Searched the [Column](https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.Column) object and didn't find an option for a prefix. I tried the following:

```python
class Users(Base, CommonMixin, TimestampMixin):
    __table_args__ = (Index('idx_search', "first_name", "last_name", "email", mysql_prefix='FULLTEXT'), )

    first_name: str = Column(String(20))
    last_name: str = Column(String(20))
    email: EmailStr = Column(String(50))
```

Running this snippet returns the following error:

```sql
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1706, "Feature 'Adding Fulltext indexes' is not supported by SingleStore.")
```

Is this the correct way to use it in declarative ORM? From the error message, it seems like it is not possible to add this prefix in SingleStore.

**Reply by zzzeek** [reply]
OK so that's a SIngleStore limitation.  we don't have a built in for that, so you'd have to make one, using [compiles](https://docs.sqlalchemy.org/en/14/core/compiler.html).   this example is not really described there as it's unusual, so it will work if you subclass Constraint as below

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.orm import declarative_base
from sqlalchemy.schema import Constraint
from sqlalchemy.schema import CreateTable


Base = declarative_base()


class Fulltext(Constraint):
    def __init__(self, arg):
        self.arg = arg
        super().__init__()


@compiles(Fulltext)
def compile_ft(elem, compiler, **kw):
    return f"FULLTEXT({elem.arg})"

class Users(Base):
    __tablename__ = 'user'

    __table_args__ = (
        Fulltext("first_name"),
    )

    id = Column(Integer, primary_key=True)
    first_name: str = Column(String(20))
    last_name: str = Column(String(20))
    email: str = Column(String(50))

print(CreateTable(Users.__table__))
```

**Reply by wiseaidev** [reply]
Found another limitation:

```python
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1706, "Forwarding Error (node-XXX): Feature 'more than one FULLTEXT KEY' is not supported by SingleStore.")
```

When running:

```python
class Users(Base, CommonMixin, TimestampMixin):
    __table_args__ = (
        Fulltext("first_name"), Fulltext("last_name"),  Fulltext("email")
    )

    first_name: str = Column(String(20))
    last_name: str = Column(String(20))
    email: EmailStr = Column(String(50))
```

I appreciate your help so much.

**Reply by zzzeek** [reply]
why dont you make it one FULLTEXT clause?   `FullText("first_name, last_name, email")`, that looked like the syntax it wanted

**Reply by wiseaidev** [reply]
That works, however, when querying on these columns, only the `first_name` column is doing the full search:

```sql
            AND
              MATCH (
                first_name,
                last_name,
                email
              )
              AGAINST (
                :search,
              )
```

**Reply by zzzeek** [reply]
So I can show you how to emit whatever DDL you want but you have to interface with people that know SingleStore in order to know how to configure the database to do the things you need.
